Vim insert setline undo fixes

eraserhd opened the pull request:
These two patches from Vim fixed eraserhd/parinfer-rust#14 .

The first prevents corruption of undo history when setline() is called from a TextChangeI handler.  The second merges undo in insert mode to behave more like the user expects.

eraserhd committed:
vim-patch:8.1.0245: calling setline() in TextChangedI autocmd breaks undo

Problem:    Calling setline() in TextChangedI autocmd breaks undo. (Jason
            Felice)
Solution:   Don't save lines for undo when already saved. (closes vim/vim#3291)
https://github.com/vim/vim/commit/91d2e783b41ca900bc603b3cb5e083c8a4a33170

eraserhd committed:
vim-patch:8.1.0256: using setline() in TextChangedI splits undo

Problem:    Using setline() in TextChangedI splits undo.
Solution:   Use another solution for undo not working properly.
https://github.com/vim/vim/commit/9fa9506853516c82851baec643aa47458cb8b3bc

This deviates from Vim in the handling of the CursorHoldI event.  In
Vim, any buffer changes are merged into the insert.  In Neovim,
CursorHoldI is handled via the multiqueue, and the point at which
the cursor hold is implemented (in input.c) doesn't know enough about it.
Making all queued events merge into the insert seems more wronger since
changes by other asynchronous events really should be separately
undoable.

justinmk commented on src/nvim/edit.c:
```diff
@@ -6117,6 +6125,7 @@ void free_last_insert(void)
 }
 
 #endif
+static int ins_apply_autocmds(event_T event);
```
forward declarations aren't needed, it's done automatically.

justinmk commented on src/nvim/edit.c:
```diff
@@ -8673,6 +8682,24 @@ static char_u *do_insert_char_pre(int c)
 
+/*
+ * Trigger "event" and take care of fixing undo.
+ */
```
/// Trigger "event" and take care of fixing undo.

justinmk commented on src/nvim/edit.c:
```diff
@@ -8673,6 +8682,24 @@ static char_u *do_insert_char_pre(int c)
+  // If u_savesub() was called then we are not prepared to start
+  // a new line.  Call u_save() with no contents to fix that.
+  if (tick != buf_get_changedtick(curbuf))
+    u_save(curwin->w_cursor.lnum, (linenr_T)(curwin->w_cursor.lnum + 1));
```
lint will want braces : 

```
if (..) {
  ..
}
```

eraserhd committed:
Fix lint

janlazo commented:
Do any these 8.1.x patches depend on earlier 8.0.x patches?

eraserhd commented:
@janlazo Nope.

I can demonstrate the issue is fixed with manual testing (the test case doesn't work because of missing `test_override()`).  I didn't bring any other code in, and it looked like the only surrounding changes were there from Neovim prior (e.g. `buf_get_changedtick(curbuf)` instead of `CHANGEDTICK(curbuf)` and formatting).

eraserhd commented:
Ugh... OK, I tested a simple case, but this might not be working yet.  DO NOT MERGE.

janlazo commented:
Did you check the blame for each file modified in the patch?
You're missing 8.1.0057.

eraserhd commented:
@janlazo Oh, thanks.  I actually notice that I missed parts of the merge.  Maybe they weren't in the `.rej` files?  I'll check that, too.

eraserhd committed:
vim-patch:8.1.0057: popup menu displayed wrong when using autocmd

Problem:    Popup menu displayed wrong when using autocmd.
Solution:   Use aucmd_prepbuf().  Force updating status line if the popup menu
            is going to be redrawn anyway. (Christian Brabandt, closes vim/vim#3009)
https://github.com/vim/vim/commit/6ba3ec1bace67513a352326864cebc16b3c5bc56

eraserhd commented:
OK, this necessarily deviates slightly from Vim:  Text from the relevant commit:

This deviates from Vim in the handling of the CursorHoldI event.  In
Vim, any buffer changes are merged into the insert.  In Neovim,
CursorHoldI is handled via the multiqueue, and the point at which
the cursor hold is implemented (in input.c) doesn't know enough about it.
Making all queued events merge into the insert seems more wronger since
changes by other asynchronous events really should be separately
undoable.

eraserhd commented:
@janlazo @justinmk I think this looks good now.  See the comment above.  The Travis CI failure looks like a fluke, and everything else passes.

janlazo commented:
@eraserhd Can you separate the changes from patch 8.1.0256 and your proposed deviation? It makes it easier to bisect and review. nvim is behind on autocmd and undo patches.

eraserhd commented:
My current deviation is just omitting a patch segment that won't apply
(around EVENT_CURSORHOLDI Which no longer exists in Neovim), so there's
nothing to separate out.

(Doing something different would require adding code in input.c or edit.c)

On Thu, Aug 9, 2018 at 7:16 PM Jan Edmund Lazo <notifications@github.com>
wrote:

> @eraserhd <https://github.com/eraserhd> Can you separate the changes from
> patch 8.1.0256 and your proposed deviation? It makes it easier to bisect
> and review. nvim is behind on autocmd and undo patches.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/8832#issuecomment-411954701>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAI_VOzACCWCmJPs194Yzr-wxfGhw0X7ks5uPO0CgaJpZM4V014b>
> .
>

eraserhd commented:
*bump* Anything I can do to move this along?

janlazo commented:
Wait for @justinmk or someone else to review again or add another vim-patch related to your changes (ie. undo) to test on current master.

justinmk commented:
@eraserhd good choices, and thanks for including the explanation in the commit message.
